### PR TITLE
Criteria query

### PR DESCRIPTION
**Pull request by Stipaxa-hub**

### CODE REVIEW

File: src/main/java/mate/project/store/repository/book/BookSpecificationProviderManager.java [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+
+@RequiredArgsConstructor
+@Component
+public class BookSpecificationProviderManager implements SpecificationProviderManager<Book> {
+    private final List<SpecificationProvider<Book>> bookSpecificationProviders;
+
+    @Override
+    public SpecificationProvider<Book> getSpecificationProvider(String key) {
+        return bookSpecificationProviders.stream()
+                .filter(p -> p.getKey().equals(key))
```
Comment (okuzan) [inline src/main/java/mate/project/store/repository/book/BookSpecificationProviderManager.java]: p - bad naming

File: src/main/java/mate/project/store/repository/book/BookSpecificationProviderManager.java [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+                .findFirst()
+                .orElseThrow(() -> new RuntimeException(
```
Comment (okuzan) [inline src/main/java/mate/project/store/repository/book/BookSpecificationProviderManager.java]: entity not found exception

File: src/main/java/mate/project/store/repository/book/spec/AuthorSpecificationProvider.java [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+import mate.project.store.entity.Book;
+import mate.project.store.repository.SpecificationProvider;
+import org.springframework.data.jpa.domain.Specification;
+import org.springframework.stereotype.Component;
+
+@Component
+public class AuthorSpecificationProvider implements SpecificationProvider<Book> {
+    @Override
+    public String getKey() {
+        return "authors";
```
Comment (okuzan) [inline src/main/java/mate/project/store/repository/book/spec/AuthorSpecificationProvider.java]: extract constants everywhere

File: src/main/resources/db/changelog/db.changelog-master.yaml [hunk 4]
```diff
@@ -0,0 +1,3 @@
+databaseChangeLog:
+  - include:
+      file: db/changelog/changes/01-create-books-table.yaml
```
Comment (okuzan) [inline src/main/resources/db/changelog/db.changelog-master.yaml]: add newline where needed

File: src/main/java/mate/project/store/repository/book/BookSpecificationBuilder.java [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+public class BookSpecificationBuilder implements SpecificationBuilder<Book> {
+    private final SpecificationProviderManager<Book> bookSpecificationProviderManager;
+
+    @Override
+    public Specification<Book> build(BookSearchParametersDto searchParametersDto) {
+        Specification<Book> spec = Specification.where(null);
+        if (searchParametersDto.titles() != null && searchParametersDto.titles().length > 0) {
+            spec = spec.and(bookSpecificationProviderManager
+                    .getSpecificationProvider("titles")
+                    .getSpecification(searchParametersDto.titles()));
```
Comment (okuzan) [inline src/main/java/mate/project/store/repository/book/BookSpecificationBuilder.java:24]: constant

File: src/main/java/mate/project/store/controller/BookController.java [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.ResponseStatus;
+import org.springframework.web.bind.annotation.RestController;
+
+@RequiredArgsConstructor
+@RestController
+@RequestMapping(value = "/books")
+public class BookController {
+
```
Comment (okuzan) [inline src/main/java/mate/project/store/controller/BookController.java]: no empty lines after class declaration

**Review by okuzan — CHANGES_REQUESTED** [review]

**Review by kozhukhovsky — APPROVED** [review]

**Pull request merged**